client/src/App.tsx

```diff
@@ -1,38 +1,195 @@
+
+//If the currentRoundRef === game.totalRounds then the game ends and a winner is declared
+//If the curerntRoundRef !== game.totalRounds then startRound() runs and to start the round for the next player
+
 function App() {
   const [game, setGame] = useState<Game>(
     initialiseGame(DEFAULT_PLAYERS, DEFAULT_ROUNDS)
   );
   const player1 = game.players[0];
   const player2 = game.players[1];
 
+  const { players, totalRounds, roundWinners } = game;
+
+  const currentRoundRef = useRef<number>(0);
+  const currentPlayerRef = useRef<Player>(players[0]);
+
+  const showPlayerCard = (player: Player) => {
+    setGame((prevGame: Game) => {
+      return {
+        ...prevGame,
+        players: updatePlayerIsCardShown(prevGame.players, player.id),
+      };
+    });
+  };
+
+  const showAllCards = () => {
+    setGame((prevGame: Game) => {
+      return {
+        ...prevGame,
+        players: updatePlayerIsCardShownAll(prevGame.players, true),
+      };
+    });
+  };
+
+  const hideAllCards = () => {
+    setGame((prevGame: Game) => {
+      return {
+        ...prevGame,
+        players: updatePlayerIsCardShownAll(prevGame.players, false),
+      };
+    });
+  };
+
+  const nextPlayer = () => {
+    const prevPlayer = currentPlayerRef.current;
+    const prevPlayerIndex = players.findIndex(
+      (player) => player.id === prevPlayer.id
+    );
+    let nextPlayer: Player;
+    if (prevPlayerIndex === players.length - 1) {
+      nextPlayer = players[0];
+    } else {
+      nextPlayer = players[prevPlayerIndex + 1];
+    }
+    currentPlayerRef.current = nextPlayer;
+    startRound(nextPlayer);
+  };
+
+  const startRound = (player: Player) => {
+    currentRoundRef.current = currentRoundRef.current + 1;
+
+    setGame((prevGame: Game) => {
+      return {
+        ...prevGame,
+        currentRound: prevGame.currentRound + 1,
+      };
+    });
+
+    showPlayerCard(player);
+
+    if (!player.isHuman) {
+      const stat = chooseRandomStat(
+        player.cards[0].stats,
+        player.cards[0].stats.length
+      );
+      setTimeout(() => {
+        playRound(stat, player);
+      }, DEFAULT_TIMEOUT);
+    }
+  };
+
+  const playRound = (stat: Stat, player: Player) => {
+    showAllCards();
+
+    setTimeout(() => {
+      const { id } = calculateRoundWinner(players, stat, player);
+      console.log("winner " + id);
+
+      setGame((prevGame: Game) => {
+        return {
+          ...prevGame,
+          players: updatePlayerScores(prevGame.players, id),
+          roundWinners: updateRoundWinners(prevGame.roundWinners, id),
+          //roundsPlayed: prevGame.roundsPlayed + 1,
+        };
+      });
+
+      endRound();
+    }, DEFAULT_TIMEOUT);
+  };
+
+  const endRound = () => {
+    hideAllCards();
+
+    setTimeout(() => {
+      setGame((prevGame: Game) => {
+        return { ...prevGame, players: updatePlayerCards(prevGame.players) };
+      });
+      if (currentRoundRef.current === totalRounds) {
```

Comment:
Here could we use game.roundWinners.length instead of currentRoundRef.current?

Reply:
See comment above, it didn't work when using currentRound from state. I assume the same would be true of roundWinners.length because they're held on the same state object. It's an issue with stale state values and when React actually re-renders.